### ISSUE TITLE
Using custom COCO dataset to train model

### ISSUE THREAD

**Issue by RPalmr** [ultralytics/ultralytics#9161]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, is it possible to train a YoloV8 model on a custom COCO dataset in this format? - 

`{
  "images": [
    {
      "id": 13004,
      "license": 1,
      "file_name": "/path/to/image1.jpg",
      "height": 2168,
      "width": 3848,
      "date_captured": "2020-11-20T08:21:10.334962+0000"
    },
    {
      "id": 34227,
      "license": 1,
      "file_name": "/path/to/image2.jpg",
      "height": 2168,
      "width": 3848,
      "date_captured": "2020-12-09T06:23:07.440353+0000"
    },
    {
      "id": 40125,
      "license": 1,
      "file_name": "/path/to/image3.jpg",
      "height": 2168,
      "width": 3848,
      "date_captured": "2021-04-21T10:45:53.557586+0000"
    },
    {
      "id": 94219,
      "license": 1,
      "file_name": "/path/to/image4.jpg",
      "height": 2168,
      "width": 3848,
      "date_captured": "2020-02-18T09:09:10.669984+0000"
    },`
    
    
 Or are there any scripts to convert this to Yolo txt format please? I have tried JSON2YOLO but I am not having any luck.

**Comment by glenn-jocher** [comment]
@RPalmr hello! 👋 Yes, you can definitely train a YOLOv8 model on a custom COCO dataset. The format you've shown is the standard COCO format for the images section. YOLOv8 can automatically handle this format during training by specifying the correct paths in your dataset YAML file. 

For your COCO dataset, ensure you have a YAML file like this:

```yaml
# dataset.yaml
train: /path/to/your/train/images  # directory to train images
val: /path/to/your/val/images  # directory to val images

# COCO json annotation files
train_ann: /path/to/your/train/annotations.json
val_ann: /path/to/your/val/annotations.json

nc: 80  # number of classes
names: [ 'class1', 'class2', ..., 'class80']  # class names
```

Then you can train your model using:

```bash
yolo train model=yolov8n.pt data=dataset.yaml
```

No need to convert to YOLO txt format for training, as YOLOv8 directly supports COCO JSON format. 🎉 If you encounter issues with JSON2YOLO or have specific needs not met by direct COCO JSON training, could you share more details so we can help further? 

Happy training!

**Comment by RPalmr** [comment]
Thank you so much Glenn. Have a great day.

**Comment by RPalmr** [comment]
Hey Glenn, thanks for your reply. My dataset.yaml folder is set exactly like the example you gave. I have extracted the images  as .jpg format into train and val folders respectively. I have set the train_ann and val_ann as the coco file. Though, when I run `yolo train model=yolov8n.pt data=dataset.yaml`, the training images are all recognised as backgrounds by yolo. Does this mean that yolo recognises the images as un-annotated?

**Comment by glenn-jocher** [comment]
@RPalmr hey there! 🌟 It sounds like you've set up your `dataset.yaml` correctly. If YOLO is recognizing all images as backgrounds, it likely means it's not correctly parsing your annotations.

For COCO datasets, it's perfectly fine to have all annotations in a single file. YOLOv8 should automatically handle this as long as the paths in your `dataset.yaml` point correctly to where your single COCO formatted annotation file for both training and validation is located.

Please double-check the `train_ann` and `val_ann` paths in your `dataset.yaml` file to ensure they accurately point to your COCO annotation file. If your training and validation images are sharing the same annotation file, make sure the IDs in the dataset correctly match those specified in the annotation.

```yaml
train_ann: /path/to/your/annotations.json
val_ann: /path/to/your/annotations.json
```

Remember, the dataset loader will filter out the relevant annotations for each image based on the `image_id` provided in the COCO JSON. If the paths and IDs are correctly set, YOLOv8 will properly recognize and use those annotations rather than defaulting to treating all images as having no annotations. 

Let me know if this helps or if you're still encountering issues!

**Comment by JosueDavalos** [comment]
Hi @glenn-jocher 

I am experiencing the same problem. I want to use my annotations in COCO format and I see that all images are detected as **background** and the loss and performance of the model is 0.
 
I am trying to train an intance segmentation model that detects only one class using google colab. Can you help me to detect and solve my problem. Here I attach details of the configuration I have. As you will see, for practical purposes, I am using the same dataset to train and validate.

I also tried adding the full path to the images (for example: from  ` "file_name": "1.jpg"` to ` "file_name": "/content/drive/MyDrive/folder_1/project_folder/images/1.jpg"`), thinking this was the problem, but I still get the same result


```json
{
  "images": [
    {
      "license": 0,
      "file_name": "1.jpg",
      "coco_url": "n/a",
      "height": 1520,
      "width": 2688,
      "date_captured": "",
      "flickr_url": "n/a",
      "id": 1149177037,
      "tag_ids": []
    },
    {
      "license": 0,
      "file_name": "2.jpg",
      "coco_url": "n/a",
      "height": 1520,
      "width": 2688,
      "date_captured": "",
      "flickr_url": "n/a",
      "id": 478292595,
      "tag_ids": []
    },
    ......
  ],
  "annotations": [
    {
      "id": 1,
      "image_id": 478292595,
      "category_id": 886886774,
      "segmentation": [[1367.0, 496.0, 1379.0]],
      "area": 24697.50390111655,
      "bbox": [1199.0, 496.0, 217.0, 237.0],
      "iscrowd": 0,
      "extra": {}
    },
    {
      "id": 2,
      "image_id": 478292595,
      "category_id": 886886774,
      "segmentation": {
        "counts": [2349218, 18, 1498],
        "size": [1520, 2688]
      },
      "area": 89721.0,
      "bbox": [1545, 554, 286, 642],
      "iscrowd": 1,
      "extra": {}
    },
    ......
  ]
}

```

```yaml
train: /content/drive/MyDrive/folder_1/project_folder/images
val: /content/drive/MyDrive/folder_1/project_folder/images

# COCO json annotation files
train_ann: /content/drive/MyDrive/folder_1/project_folder/labels.json
val_ann: /content/drive/MyDrive/folder_1/project_folder/labels.json

nc: 1
names: ["class_name"]
```

```python
model = YOLO('yolov8n-seg.pt')

dataset_path = '/content/drive/MyDrive/folder_1/project_folder/security-cam.yaml'
results = model.train(
    data=dataset_path,
    project="people-tracking",
    name="baseline",
    single_cls=True,
    plots=True,
    epochs=2,
    imgsz=640,
)
```

**Comment by glenn-jocher** [comment]
Hi there! 👋 It seems like the issue might relate to how the annotations are linked to the images. Here are a few checks and tips that might help:

1. **Check Image IDs:** Ensure that the `image_id` in your annotations correctly matches the `id` in your images section of the JSON. This is crucial for the model to correctly map annotations to images.

2. **Annotation Formats:** For instance segmentation, ensure that your segmentation formats are consistent and correctly formatted. Sometimes discrepancies in format (like mixing polygon and RLE) could cause issues.

3. **Paths in COCO JSON:** Since you adjusted the `file_name` to include the full path, make sure this path is exactly correct and accessible in your Google Colab environment.

4. **Single Class Configuration:** Since you're training for a single class, setting `single_cls=True` is correct. However, ensure your `category_id` in annotations is consistent and matches the class ID expected in the YAML configuration (which starts from 0 for COCO datasets in YOLO).

Here's a snippet to check if your annotation IDs match your image IDs from within Colab, which can help debugging:

```python
import json

# Load your annotations json
with open('/content/drive/MyDrive/folder_1/project_folder/labels.json') as f:
    data = json.load(f)
    
# Check for unmatched IDs
image_ids = {img['id'] for img in data['images']}
annotation_image_ids = {ann['image_id'] for ann in data['annotations']}

# Unmatched IDs
unmatched_ids = annotation_image_ids - image_ids
print("Unmatched IDs:", unmatched_ids)  # Should be empty

# This will help ensure that every image_id referenced in annotations is declared in images
```

Run this snippet to verify if all `image_id` values in annotations are declared in images. If there are unmatched IDs, it explains why they are treated as background (which means no annotations found for those IDs).

Hope this helps! Let me know how it goes or if there's anything else you'd like to explore. 🚀

**Comment by ZeeRizvee** [comment]
I am facing the same issue. This might be a bug in the conversion of JSON labels to TXT format.

**Comment by glenn-jocher** [comment]
Hello @ZeeRizvee,

Thank you for reaching out! It sounds like you're encountering an issue where your images are being recognized as backgrounds during training, which might be due to a problem with the conversion of JSON labels to the TXT format.

To help us diagnose this issue more effectively, could you please provide a minimum reproducible example? This will allow us to better understand the context and specifics of your setup. You can find guidance on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure that you are using the latest versions of the Ultralytics and YOLOv8 packages. Sometimes, updates include important bug fixes and improvements that might resolve your issue.

If you have already verified the versions and the problem persists, please share a snippet of your JSON annotations and the corresponding YAML configuration. This will help us pinpoint any potential mismatches or formatting issues.

Looking forward to your response so we can assist you further! 😊

**Comment by alrichardbollans** [comment]
I am having the same issue where no labels are found and the images are all set to backgrounds. Using ultralytics v.8.2.74 with
```
from ultralytics import YOLO
model = YOLO("yolov8n-seg.pt")  # load a pretrained model (recommended for training)

results = model.train(
    data='dataset.yaml', single_cls=False,
    epochs=5)
    ```
 Where the YAML file is:
 ```
train: "/home/annotations work/images"
val: "/home/annotations work/images"
train_ann: "/home/annotations work/instances_default.json"
val_ann: "/home/annotations work/coco_polygons/annotations/instances_default.json"
nc: 2
names: ['Viable', 'Non viable']
```

And the coco annotations are: [instances_default_MWE.json](https://github.com/user-attachments/files/16548958/instances_default_MWE.json)

**Comment by pderrenger** [comment]
Thank you for providing the details. It seems like there might be an issue with how the annotations are being parsed. Please ensure that the `image_id` in your annotations matches the `id` in your images section of the JSON. Also, verify that the paths in your YAML file are correct and accessible. If the issue persists, please try updating to the latest version of the Ultralytics package to see if the problem is resolved. If the problem continues, please let us know so we can further investigate.

**Comment by alrichardbollans** [comment]
Thanks for the response. I've checked the `image_id` and `id`s match in the appropriate places and that the paths in the YAML file are correct. I've also updated ultralytics to the last version but still getting this error.

I've made a workaround for myself by converting the coco annotations into YOLO format with the ultralytics `convert_coco` method (https://docs.ultralytics.com/reference/data/converter/#ultralytics.data.converter.convert_coco), and then copying the images into the appropriate folder (something like save_dir/images/default), and finally making a yaml file following https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/#22-create-labels. With this set up, the training now runs as expected.

**Comment by pderrenger** [comment]
Thank you for the update. It's great to hear that you found a workaround by converting the COCO annotations into YOLO format using the `convert_coco` method. This approach is indeed effective and aligns well with the recommended practices for custom data training. If you encounter any further issues or have additional questions, feel free to reach out. We're here to help!

**Comment by mesllo-bc** [comment]
@glenn-jocher I'm also getting this issue with version `8.2.79`. My `yaml` file is as follows:
```
# dataset.yaml
train: /path/to/dataset/images  # directory to train images
val: /path/to/dataset/images  # directory to val images

# COCO json annotation files
train_ann: /path/to/dataset/coco_train.json
val_ann: /path/to/dataset/coco_val.json

nc: 40  # number of classes
names: [ 'class_1', ... 'class_40']
```

Note that the `file_name` path is set to something like `some_dataset_subfolder/image_name.jpg`, and I assume that the above `yaml` file will be used by the `ultralytics` package to concatenate the path into the following: `/path/to/dataset/images/some_dataset_subfolder/image_name.jpg`.

I've also ran the snippet which confirms that the `image_id` in the `annotations` and the `id` in `images` match perfectly.

Also note that:
1. The classes start from 1
2. The `bbox` is not normalized

Basically, everything is default as COCO expects it. It's my understanding that the conversion to the `YOLO` format is handled implicitly according to this:

> @RPalmr hello! 👋 Yes, you can definitely train a YOLOv8 model on a custom COCO dataset. The format you've shown is the standard COCO format for the images section. YOLOv8 can automatically handle this format during training by specifying the correct paths in your dataset YAML file.

For running the training I am attempting the following:
```
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
model.train(data="path/to/the/above/yaml/file.yaml", epochs=3)  # train the model
```

The result I am getting is:
```
...
...
18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
...
Transferred 355/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
...
train: Scanning [/path/to/dataset/labels/some_dataset_subfolder/some_image_file_name...](...)
0 images, 46330 backgrounds, 0 corrupted
```

Therefore, it seems that it is still expecting the `labels` folder to be present I'm guessing? I thought that what the above quote implies is that this is also handled. If not, could you kindly indicate what we should do, please? :) 

Thank you!

**Comment by glenn-jocher** [comment]
Thank you for the detailed information. It appears that the issue might be related to how the paths are being handled. The `file_name` in your COCO JSON should be relative to the paths specified in your `dataset.yaml`. Please ensure that the paths are correctly concatenated. If the issue persists, consider using the `convert_coco` method to convert your annotations to YOLO format as a workaround.

**Comment by margajdon** [comment]
> Thank you for the detailed information. It appears that the issue might be related to how the paths are being handled. The `file_name` in your COCO JSON should be relative to the paths specified in your `dataset.yaml`. Please ensure that the paths are correctly concatenated. If the issue persists, consider using the `convert_coco` method to convert your annotations to YOLO format as a workaround.

I've been having this same issue, tried this fix and it didn't work for me. For reference, my dataset.yaml file:
```
train: path/to/data/train/images
val: path/to/data/test/images

train_ann: path/to/data/train/annotations.json
val_ann: path/to/data/test/annotations.json

nc: 3

names: [c1, c2, c3]
```

In my COCO annotations.json file for the training split, I've now tried for filenames `path/to/data/train/images/filename.jpg`, `train/images/filename.jpg`, `images/filename.jpg` and `filename.jpg` and for each option I still get everything recognized as background. Am I maybe misunderstanding how exactly the filenames should be concatenated?

**Comment by mesllo-bc** [comment]
Thanks for your reply @glenn-jocher!

I can confirm that the filenames are concatenated correctly. The problem seems to happen in `cache_labels()` of the `YOLODataset` in line 134 of `data/dataset.py`.

When it tries to verify the image labels via `verify_image_label` in `data/utils.py`, it loads the following args:
```
('/path/to/dataset/images/example_subfolder/example_image.jpg', '/path/to/dataset/labels/example_subfolder/example_image.txt', '\x1b[34m\x1b[1mtrain: \x1b[0m', False, 40, 0, 0)
```

Indeed, it seems that it already expects some kind of predefined YOLO format here, where it expects a `.txt` file-based structure by default. However, earlier in this post (as per the below quote), it was stated that we don't need to do this as it will be automatically handled:

> @RPalmr hello! 👋 Yes, you can definitely train a YOLOv8 model on a custom COCO dataset. The format you've shown is the standard COCO format for the images section. YOLOv8 can automatically handle this format during training by specifying the correct paths in your dataset YAML file.

Is there something I'm missing here, maybe some flag that needs to be set? Or is the `convert_coco` method the only actual way to do this?

**Comment by glenn-jocher** [comment]
Thank you for the detailed follow-up. It appears that the issue might be related to how the labels are being processed. While YOLOv8 should handle COCO format annotations, you might need to ensure that the paths and formats are correctly set. If the problem persists, using the `convert_coco` method to convert your annotations to YOLO format is a reliable workaround. This method ensures compatibility and might resolve the issue you're facing.

**Comment by glenn-jocher** [comment]
Thank you for the clarification. If the paths and labels are correct and the issue persists, it might be a bug. Please ensure you're using the latest version of the Ultralytics package. If the problem continues, consider opening a new issue on GitHub with detailed steps to reproduce the problem.

**Comment by BenedictWilkins** [comment]
@glenn-jocher this also doesn't work for me, I have a similar set up to @margajdon, have followed all the guidance here. My dataset loads properly when using  `torchvision.datasets.CocoDetection` (including segmentation data) so no bad coco formatting issues - I am using latest pip versions of things (fresh colab install ultralytics==8.2.84), may be a bug?

**Comment by glenn-jocher** [comment]
@BenedictWilkins thank you for the details. Since you're using the latest version and the issue persists, it might indeed be a bug. Please open a new issue on GitHub with detailed steps to reproduce the problem, and we'll investigate further.

**Comment by glenn-jocher** [comment]
If your dataset works with `torchvision.datasets.CocoDetection` but not with YOLOv8, it might be a bug. Please consider opening an issue on GitHub with detailed steps to reproduce the problem.

**Comment by glenn-jocher** [comment]
If your dataset works with `torchvision.datasets.CocoDetection` but not with YOLOv8, it might be a bug. Please consider opening an issue on GitHub with details to help us investigate further.

**Comment by Gnoot01** [comment]
if YOLO works directly with coco json files, then why is JSON2YOLO a thing? @glenn-jocher

**Comment by glenn-jocher** [comment]
If your setup works with `torchvision.datasets.CocoDetection` but not with Ultralytics, it might be a bug. Please consider opening an issue on GitHub with details for further investigation.

**Comment by glenn-jocher** [comment]
@BenedictWilkins if everything is set up correctly and the issue persists, it might be a bug. Please consider opening a new issue on GitHub with detailed steps to reproduce the problem.

**Comment by glenn-jocher** [comment]
If your dataset works with `torchvision.datasets.CocoDetection` and you're experiencing issues with Ultralytics, it might be a bug. Please consider opening a new issue on GitHub with detailed steps to reproduce the problem.

**Comment by glenn-jocher** [comment]
@Gnoot01 YOLO does not work directly with COCO JSON dataset annotations, they must first be exported to YOLO (txt) format.
